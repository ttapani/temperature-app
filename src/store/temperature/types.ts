import { Action } from 'redux';


export interface TemperatureState {
    data: Temperature[];
    isLoading: boolean;
    error?: string;
}

export interface Temperature {
    id: number;
    name: string;
    temperature: number;
    lastUpdated?: number;
    favourite: boolean;
}

export interface GetTemperatureAction extends Action {
    type: 'temperature/GET';
    payload: {
        id: number;
    };
}

export interface GetTemperatureSuccessAction extends Action {
    type: 'temperature/GET_SUCCESS';
    payload: {
        data: Temperature;
    };
}

export interface GetTemperatureCancelAction extends Action {
    type: 'temperature/GET_CANCEL';
}

export interface GetTemperatureFailureAction extends Action {
    type: 'temperature/GET_FAILURE';
    payload: {
        error: string;
    };
}

export interface TemperatureClearAction extends Action {
    type: 'temperature/CLEAR';
}

export interface UpdateFavouritesAction extends Action {
    type: 'temperature/UPDATE';
}

export interface AddFavouriteAction extends Action {
    type: 'temperature/ADD_FAVOURITE';
    payload: {
        location: Temperature;
    };
}

export interface RemoveFavouriteAction extends Action {
    type: 'temperature/REMOVE_FAVOURITE';
    payload: {
        location: Temperature;
    };
}

export type TemperatureDataAction = GetTemperatureAction |Â GetTemperatureSuccessAction | GetTemperatureCancelAction | TemperatureClearAction | GetTemperatureFailureAction;
export type TemperatureFavouriteAction = AddFavouriteAction | RemoveFavouriteAction | UpdateFavouritesAction;
export type TemperatureAction = TemperatureDataAction | TemperatureFavouriteAction;