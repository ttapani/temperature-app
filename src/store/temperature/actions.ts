import { ActionCreator } from 'redux';
import { Temperature, TemperatureState, TemperatureDataAction } from './types';
import { GetTemperatureAction, GetTemperatureSuccessAction, GetTemperatureCancelAction, GetTemperatureFailureAction, TemperatureClearAction } from './types';
import { ToggleFavouriteAction, UpdateFavouritesAction } from './types';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import moment from 'moment';


export const getTemperatureStarted: ActionCreator<GetTemperatureAction> = (id: number) => ({
    type: 'temperature/GET',
    payload: {
        id,
    },
});

export const getTemperatureSuccess: ActionCreator<GetTemperatureSuccessAction> = (data: Temperature) => ({
    type: 'temperature/GET_SUCCESS',
    payload: {
        data,
    },
});

export const getTemperatureCancel: ActionCreator<GetTemperatureCancelAction> = () => ({
    type: 'temperature/GET_CANCEL',
});

export const getTemperatureFailure: ActionCreator<GetTemperatureFailureAction> = (error: string) => ({
    type: 'temperature/GET_FAILURE',
    payload: {
        error,
    },
});

export const clearTemperature: ActionCreator<TemperatureClearAction> = () => ({
    type: 'temperature/CLEAR',
});

export const toggleFavourite: ActionCreator<ToggleFavouriteAction> = (id: number) => ({
    type: 'temperature/TOGGLE_FAVOURITE',
    payload: {
        id,
    },
});

export interface WeatherData {
    name: string;
    id: number;
    main: {
        temp: number;
    }
}

export const getTemperature = (id: string, update?: boolean,cancelRequest?: () => boolean): ThunkAction<Promise<void>, TemperatureState, {}, TemperatureDataAction> => {
    return async (dispatch: ThunkDispatch<TemperatureState, {}, TemperatureDataAction>, getState): Promise<void> => {
        return new Promise<void>((resolve, reject) => {
            dispatch(getTemperatureStarted(id))
            const response = fetch(`${process.env.REACT_APP_REQUEST_URL}?q=${id}&APPID=${process.env.REACT_APP_API_KEY}&units=metric`);
            response.then((response) => {
                if(response.status === 200) {
                    return response.json();
                } else {
                    dispatch(getTemperatureFailure());
                    resolve();
                }
            }).then((data: WeatherData) => {
                let cancelled = false;
                if(cancelRequest) {
                    cancelled = cancelRequest();
                }
                if(data && !cancelled) {
                    dispatch(getTemperatureSuccess({ id: data.id, name: data.name, temperature: data.main.temp, lastUpdated: Date.now(), favourite: update }));
                } else {
                    dispatch(getTemperatureCancel());
                }
                resolve();
            }).catch((error) => {
                dispatch(getTemperatureFailure(error));
                resolve();
            });
        });
    };
};

export const updateFavourites = (): ThunkAction<Promise<void>, TemperatureState, {}, UpdateFavouritesAction> => {
    return async (dispatch: ThunkDispatch<TemperatureState, {}, UpdateFavouritesAction>, getState): Promise<void> => {
        return new Promise<void>((resolve, reject) => {
            let state = getState();
            state.data.forEach(location => {
                if(moment(location.lastUpdated).add(1, 'm').isBefore(moment())) {
                    dispatch(getTemperature(location.name, true));
                }
            })
        });
    };
};