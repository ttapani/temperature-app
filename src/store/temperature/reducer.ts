import { Reducer } from 'redux';
import { TemperatureState, TemperatureAction, Temperature } from './types'; 


export const initialState: TemperatureState = {
    data: [],
    favourites: [],
    isLoading: false,
};

const temperatureReducer: Reducer<TemperatureState> = (state: TemperatureState = initialState, action): TemperatureState => {
    switch ((action as TemperatureAction).type) {
        case 'temperature/GET':
            return { ...state, isLoading: true };
        case 'temperature/GET_SUCCESS':
            return { ...state, data: updateOrAddObjectInArray(state.data, action.payload.data), isLoading: false };
        case 'temperature/GET_CANCEL':
            return { ...state, isLoading: false };
        case 'temperature/CLEAR':
            return { ...state, data: [] };
        case 'temperature/GET_FAILURE':
            return { ...state, error: action.payload.error, isLoading: false };
        case 'temperature/ADD_FAVOURITE':
            return { ...state, favourites: [...state.favourites, action.payload.location] };
        case 'temperature/REMOVE_FAVOURITE':
            return { ...state, favourites: state.favourites.filter(favourite => favourite.id !== action.payload.location.id) };
        default:    
            return state;
        }
};
    
// This would be a cool case to use generics, but all our types would have to extend a type that always has id..
const updateOrAddObjectInArray = (array: Temperature[], updatedItem: Temperature) => {
    const index = array.findIndex(item => item.id === updatedItem.id);
    if(index > -1) {
        let newArray = array.slice(0, index);
        newArray = newArray.concat(updatedItem);
        newArray = newArray.concat(array.slice(index + 1, array.length));
        return newArray;
    }
    else {
        return [...array, updatedItem];
    }
};


export default temperatureReducer;