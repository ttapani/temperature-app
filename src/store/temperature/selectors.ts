import { createSelector } from 'reselect'
import { TemperatureState, Temperature } from './types';


const getData = (state: TemperatureState) => state.data;

export const getFavouriteLocations = createSelector(
    [getData],
        (data: Temperature[]) => {
            return data.filter(loc => loc.favourite);
        }
)

export const getSearchResultLocations = createSelector(
    [getData],
        (data: Temperature[]) => {
            return data.filter(loc => !loc.favourite);
        }
)