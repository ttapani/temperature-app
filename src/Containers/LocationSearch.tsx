import React, { Component, createRef } from 'react';
import { Button, Paper, IconButton, InputBase, createStyles, Theme, withStyles, WithStyles } from '@material-ui/core';
import { Temperature, TemperatureState, TemperatureAction } from '../store/temperature/types';
import { getTemperature } from '../store/temperature/actions';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import SearchIcon from '@material-ui/icons/Search';
import CancelIcon from '@material-ui/icons/Cancel';


interface IProps extends WithStyles<typeof styles> {
    onChangeHandler: (value: string) => void;
    onCancelSearch: () => void;
}

interface IStateProps {
    data: Temperature[];
}

interface IDispatchProps {
    getTemperature: (id: string, cancelRequest?: () => boolean) => void;
}

interface IState {
    inputValue: string;
}

type Props = IStateProps & IProps & IDispatchProps;


export const styles = (theme: Theme) => createStyles({
    root: {
        padding: '2px 4px',
        display: 'flex',
        alignItems: 'center',
        width: '95%',
    },
    input: {
        marginLeft: 8,
        flex: 1,
    },
    iconButton: {
        padding: 10,
    },
});


export class LocationSearch extends Component<Props, IState> {
    constructor(props: Props) {
        super(props);
        this.state = { inputValue: '' };
    }

    handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        const value = event.target.value;
        this.setState({ inputValue: value });
        this.props.onChangeHandler(value);
    }

    cancelSearch = (event: React.MouseEvent<HTMLButtonElement>) => {
        this.setState({ inputValue: '' });
        this.props.onCancelSearch();
    }

    renderCancel = () => {
        const { classes } = this.props;
        if(this.state.inputValue !== '') {
            return (
                <IconButton className={classes.iconButton} aria-label="Cancel" onClick={this.cancelSearch}>
                    <CancelIcon />
                </IconButton>
            );
        }
    }

    render() {
        const { classes } = this.props;
        return (
            <Paper className={classes.root} elevation={1}>
                <IconButton className={classes.iconButton} aria-label="Search">
                    <SearchIcon />
                </IconButton>
                <InputBase value={this.state.inputValue} className={classes.input} placeholder="Search for locations" onChange={this.handleInputChange} />
                {this.renderCancel()}
            </Paper>
          );
    }
}

const mapStateToProps = (state: TemperatureState): IStateProps => ({ data: state.data });

const mapDispatchToProps = (dispatch: ThunkDispatch<TemperatureState, {}, TemperatureAction>): IDispatchProps => {
    return {
        getTemperature: async(id: string, cancelRequest?: () => boolean) => {
            await dispatch(getTemperature(id, cancelRequest))
        }
    }
}

export const StyledLocationSearch = withStyles(styles)(LocationSearch);
export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(LocationSearch));