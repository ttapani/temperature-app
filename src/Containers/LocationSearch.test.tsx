import React from 'react';
import { StyledLocationSearch } from './LocationSearch';
import renderer from 'react-test-renderer';
import { Temperature } from '../store/temperature/types';
import { MuiThemeProvider } from '@material-ui/core';
import { createMuiTheme } from '@material-ui/core/styles';
import red from '@material-ui/core/colors/red';

const theme = createMuiTheme({
    palette: {
        primary: red,
    },
    typography: {
        useNextVariants: true,
    },
});

test('Renders a search bar', () => {
    const locations: Temperature[] = [ { id: 1, name: 'Asd', temperature: 5 }, { id: 1, name: 'Asd', temperature: 5 }];
    const component = renderer.create(
        <MuiThemeProvider theme={theme}>
            <StyledLocationSearch data={locations} getTemperature={() => null} onChangeHandler={() => null} onCancelSearch={() => null} />
        </MuiThemeProvider>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });