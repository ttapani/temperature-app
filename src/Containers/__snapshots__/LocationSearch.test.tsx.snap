// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a search bar 1`] = `
<div
  className="MuiPaper-root-4 MuiPaper-elevation1-7 MuiPaper-rounded-5 LocationSearch-root-1"
>
  <button
    aria-label="Search"
    className="MuiButtonBase-root-37 MuiIconButton-root-31 LocationSearch-iconButton-3"
    disabled={false}
    onBlur={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiIconButton-label-36"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-40"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root-66"
    />
  </button>
  <div
    className="MuiInputBase-root-49 LocationSearch-input-2"
    onClick={[Function]}
  >
    <input
      className="MuiInputBase-input-59"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Search for locations"
      type="text"
      value=""
    />
  </div>
</div>
`;
